### ISSUE TITLE
hird_party/abseil-cpp does not contain a CMakeLists.txt file.

### ISSUE THREAD

**Issue by zoujiaqing** [protocolbuffers/protobuf#15329]
```
zoujiaqing@mac ios-build % cd protobuf-25.1/
zoujiaqing@mac protobuf-25.1 % mkdir ios-build
zoujiaqing@mac protobuf-25.1 % cd ios-build 
zoujiaqing@mac ios-build % cmake -G Xcode -Dprotobuf_BUILD_TESTS=OFF ..
-- The C compiler identification is AppleClang 15.0.0.15000100
-- The CXX compiler identification is AppleClang 15.0.0.15000100
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- 
-- 25.1.0
-- Performing Test protobuf_HAVE_LD_VERSION_SCRIPT
-- Performing Test protobuf_HAVE_LD_VERSION_SCRIPT - Failed
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Found ZLIB: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/lib/libz.tbd (found version "1.2.12")  
-- Performing Test protobuf_HAVE_BUILTIN_ATOMICS
-- Performing Test protobuf_HAVE_BUILTIN_ATOMICS - Success
CMake Warning at cmake/abseil-cpp.cmake:28 (message):
  protobuf_ABSL_PROVIDER is "module" but ABSL_ROOT_DIR is wrong
Call Stack (most recent call first):
  CMakeLists.txt:294 (include)


CMake Error at third_party/utf8_range/CMakeLists.txt:31 (add_subdirectory):
  The source directory

    /Users/zoujiaqing/projects/privchat/protobuf-25.1/third_party/abseil-cpp

  does not contain a CMakeLists.txt file.


-- Configuring incomplete, errors occurred!
```

**Comment by acozzette** [comment]
I think you just need to populate the Git submodule in third_party/abseil-cpp. This should do the trick: `git submodule update --init --recursive`

**Comment by FreeSpirit9009** [comment]
"git submodule update --init --recursive" does not work for me, please kindly reopen.

Here are the steps I performed. Environment is Ubuntu 24.04 WSL, downloaded from ubuntu.com Jun 24, 2025.
git clone --recurse-submodules https://github.com/protocolbuffers/protobuf.git
cd protobuf
git submodule update --init --recursive
git checkout v31.1
mkdir build
cd build
// at this point, I copied in the abseil-cpp (cloned from abseil's github) directory to the thirdparty directory in protobuf
cmake ..

Here is the result:
g@dinmPF59X87W:~/localbuilds/protobuf/build$ cmake ..
-- The C compiler identification is GNU 13.3.0
-- The CXX compiler identification is GNU 13.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- protobuf version: 31.1.0
-- Performing Test protobuf_HAVE_LD_VERSION_SCRIPT
-- Performing Test protobuf_HAVE_LD_VERSION_SCRIPT - Success
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Could NOT find ZLIB (missing: ZLIB_LIBRARY ZLIB_INCLUDE_DIR)
-- Performing Test protobuf_HAVE_BUILTIN_ATOMICS
-- Performing Test protobuf_HAVE_BUILTIN_ATOMICS - Success
-- Configuring done (5.2s)
CMake Error at cmake/tests.cmake:91 (target_link_libraries):
  Target "libtest_common" links to:

    absl::scoped_mock_log

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  CMakeLists.txt:333 (include)


CMake Error at cmake/tests.cmake:109 (target_link_libraries):
  Target "tests" links to:

    absl::scoped_mock_log

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  CMakeLists.txt:333 (include)


CMake Error at cmake/tests.cmake:121 (target_link_libraries):
  Target "fake_plugin" links to:

    absl::scoped_mock_log

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  CMakeLists.txt:333 (include)


CMake Error at cmake/tests.cmake:133 (target_link_libraries):
  Target "test_plugin" links to:

    absl::scoped_mock_log

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  CMakeLists.txt:333 (include)

CMake Error at cmake/tests.cmake:157 (target_link_libraries):
  Target "lite-test" links to:

    absl::scoped_mock_log

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  CMakeLists.txt:333 (include)

CMake Error at cmake/tests.cmake:185 (target_link_libraries):
  Target "lazily-build-dependencies-test" links to:

    absl::scoped_mock_log

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  CMakeLists.txt:333 (include)

CMake Error at cmake/tests.cmake:233 (target_link_libraries):
  Target "upb-test" links to:

    absl::scoped_mock_log

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  CMakeLists.txt:333 (include)

-- Generating done (0.1s)
CMake Generate step failed.  Build files cannot be regenerated correctly.

**Comment by FreeSpirit9009** [comment]
Additional information: I was able to build protobuf 4.22.5 using the following steps:

git clone --recurse-submodules https://github.com/protocolbuffers/protobuf.git
cd protobuf
git checkout v4.22.5
git submodule update --init --recursive
mkdir build
cd build
cmake ..
make
sudo make install

I noticed that the with this tag, there is a reference to abseil-cpp in the third_party directory. In the latest tag, such reference is no longer present, so I am wondering how "git submodule update" should help.

The code generated by this protobuf installation gives linking errors; I intend to create a separate issue for that.
